src/corehost/cli/fxr/fx_ver.cpp

```diff
@@ -108,20 +115,125 @@ int fx_ver_t::compare(const fx_ver_t&a, const fx_ver_t& b)
         return (a.m_patch > b.m_patch) ? 1 : -1;
     }
 
-    if (a.m_pre.empty() != b.m_pre.empty())
+    if (a.m_pre.empty() || b.m_pre.empty())
     {
-        // Either a is empty or b is empty
-        return a.m_pre.empty() ? 1 : -1;
+        // Either a is empty or b is empty or both are empty
+        return a.m_pre.empty() ? !b.m_pre.empty() : -1;
     }
 
-    // Either both are empty or both are non-empty (may be equal)
-    int pre_cmp = a.m_pre.compare(b.m_pre);
-    if (pre_cmp != 0)
+    // Both are non-empty (may be equal)
+    size_t idStart = 1;
+    for (size_t i = idStart; true; ++i)
     {
-        return pre_cmp;
+        if (a.m_pre[i] != b.m_pre[i])
+        {
+            // Found first character with a difference
+            if (a.m_pre[i] == 0 && b.m_pre[i] == _X('.'))
+            {
+                // identifiers both complete, b has an additional ident
+                return -1;
+            }
+
+            if (b.m_pre[i] == 0 && a.m_pre[i] == _X('.'))
+            {
+                // identifiers both complete, a has an additional ident
+                return 1;
+            }
+
+            // identifiers must not be empty
+            pal::string_t ida = getId(a.m_pre, idStart);
+            pal::string_t idb = getId(b.m_pre, idStart);
+
+            unsigned idanum = 0;
+            bool idaIsNum = try_stou(ida, &idanum);
+            unsigned idbnum = 0;
+            bool idbIsNum = try_stou(idb, &idbnum);
+
+            if (idaIsNum && idbIsNum)
+            {
+                // Numeric comparison
+                return (idanum > idbnum) ? 1 : -1;
+            }
+            else if (idaIsNum || idbIsNum)
+            {
+                // Mixed compare.  Spec: Number < Text
+                return idbIsNum ? 1 : -1;
+            }
+            // Ascii compare
+            return ida.compare(idb);
+        }
+        else
+        {
+            // a.m_pre[i] == b.m_pre[i]
+            if (a.m_pre[i] == 0)
+            {
+                break;
+            }
+            if (a.m_pre[i] == _X('.'))
+            {
+                idStart = i + 1;
+            }
+        }
     }
-    
-    return a.m_build.compare(b.m_build);
+
+    return 0;
+}
+
+static bool validIdentifier(const pal::string_t& id, bool buildMeta)
+{
+    if (id.empty())
+    {
+        // Identifier must not be empty
+        return false;
+    }
+
+    if (id.find_first_not_of(_X("-0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")) != pal::string_t::npos)
```

Comment:
Perf nit: wondering if it would be faster to do range checks, it's basically one value (hyphen) and then 3 range checks per character, which is probably much faster than a linear search of a relatively long string.